CONTRIBUTING.md

```diff
@@ -0,0 +1,219 @@
+a [Pull Request][PRs] on the [BOUT-dev] page. One of the maintainers will review
+your code and merge it into **next**. They may give you comments to improve the
+code. You can make additional changes and push them to the same feature branch
+and they will be automatically added to the pull request.
+
+
+## Coding Style ##
+
+Code is read an order of magnitude more times than it is written. It's also
+written for *people* and not for the computer! For these reasons, it's important
+that we stick to some form of coding standards. The following coding style
+guidelines broadly follow
+the [LLVM Coding Standards][LLVM].
+
+These guidelines are intended to make code easier to read and therefore easier
+to understand. Being consistent in coding style also helps comprehension by
+reducing cognitive load.
+
+
+### Comments ###
+
+
+Comments in the code are vital to helping understanding. Comments that
+are embedded in the code should explain **why** something is done,
+rather than *how*. For documenting what functions and classes do, we
+use [Doxygen](www.doxygen.org).
+
+- Prefer C++ style comments `//` over C style `/* */`
+
+- Doxygen comments: use `///`.
+
+  Doxygen done right:
+
+  ```
+  /// Foo the bar
+  ///
+  /// Apply the standard foo method to @p bar
+  ///
+  /// Typical usage:
+  ///
+  ///    foo(bar, "simple", result);
+  ///
+  ///
+  /// @param[in] quux Number of times to foo
+  /// @param[out] result filled with quux fooed bars
+  ///
+  /// @returns true on success
+  bool applyFoo(BoutReal bar, int quux, std::vector<int> &result);
+  ```
+
+- The header files are essentially the "public" API, so prefer to put
+  Doxygen comments there, rather than in the implementation. "Private"
+  functions, etc., can be documented in the implementation.
+
+### Naming ###
+
+Naming things correctly is super important! It is also one of the trickiest
+parts of coding.
+
+Names should be *descriptive*. Code is read an order of magnitude more
+often than it is written, so it is vital that it is easy to comprehend.
+
+There are some conventions you should follow when naming things:
+
+- Type or class names should be nouns and be PascalCase --
+  e.g. BoutReal, BoutMesh, Laplacian
+- Variable names should be nouns and snake_case -- e.g. generator, forward_map,
+  extra_yguards_lower
+- Functions should be verbs (i.e. actions) and camelCase --
+  e.g. solve, getSection, parseString
+
+Prefer a longer descriptive name over a shorter abbreviated one:
+`inner_boundary_flags` rather than `inbndflgs`, `generator` rather than
+`gen`. It's much easier to read and comprehend than the abbreviated form.
+
+### Details ###
+
+- Use spaces instead of tabs. Tabs may be interpreted differently by different
+  editors, making the code look badly indented and difficult to read. The
+  easiest solution is just use spaces everything instead.
+- Two spaces for indentation
+- Spaces after `if`, `for`, etc.
+
+  Wrong:
```

Comment:
This will be parsed. Unless you read an md file in a txt editor, highlighting whitespace ...